File: src/main/java/io/github/foodjournal/terminal/TerminalFrontend.java
package io.github.foodjournal.terminal;

import io.github.foodjournal.application.TransientVoicePayload;
import io.github.foodjournal.messaging.Attachment;
import io.github.foodjournal.messaging.MessagingFrontend;
import java.time.Duration;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

/** Local stand-in transport for the "terminal" provider; captures replies in-process instead of calling a real API. Terminal usage is single-threaded, so no correlation id is needed between a send and its reply. */
@Component
@Profile("terminal")
public class TerminalFrontend implements MessagingFrontend {
  private final BlockingQueue<String> deliveries = new LinkedBlockingQueue<>();

  @Override public String provider() { return "terminal"; }
  @Override public boolean enabled() { return true; }
  @Override public int messageLimit() { return 4096; }
  @Override public String send(String conversationId, String text) { deliveries.add(text); return "0"; }
  @Override public void edit(String conversationId, String messageId, String text) { deliveries.add(text); }
  @Override public TransientVoicePayload download(Attachment attachment) { throw new UnsupportedOperationException("Terminal mode does not support media attachments"); }

  public String awaitReply(Duration timeout) throws InterruptedException {
    return deliveries.poll(timeout.toNanos(), TimeUnit.NANOSECONDS);
  }

  public void reset() { deliveries.clear(); }
}

File: python-app/app/terminal/terminal_frontend.py
"""Local stand-in transport for the "terminal" provider; captures replies
in-process instead of calling a real API. Terminal usage is single-threaded
and sequential (one send() awaited at a time), so a simple FIFO queue is
sufficient -- no correlation id needed between a send and its reply."""

import asyncio

from app.messaging.inbound_message import Attachment


class TerminalFrontend:
    def __init__(self) -> None:
        self._deliveries: asyncio.Queue[str] = asyncio.Queue()

    def provider(self) -> str:
        return "terminal"

    def enabled(self) -> bool:
        return True

    def message_limit(self) -> int:
        return 4096

    async def send(self, conversation_id: str, text: str) -> str:
        await self._deliveries.put(text)
        return "0"

    async def edit(self, conversation_id: str, message_id: str, text: str) -> None:
        await self._deliveries.put(text)

    async def download(self, attachment: Attachment) -> bytes:
        raise NotImplementedError("Terminal mode does not support media attachments")

    async def await_reply(self, timeout: float) -> str | None:
        try:
            return await asyncio.wait_for(self._deliveries.get(), timeout=timeout)
        except TimeoutError:
            return None

    def reset(self) -> None:
        while not self._deliveries.empty():
            self._deliveries.get_nowait()
